src/Transports.AspNetCore/WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,253 @@
+                        _ = CloseAsync();
+                        // wait until the close has been sent
+                        var completedTask = Task.WhenAny(
+                            _outputClosed.Task,
+                            Task.Delay(_closeTimeoutMs, RequestAborted));
+                    }
+                    // quit
+                    return;
+                }
+                // if this is the last block terminating a message
+                if (result.EndOfMessage)
+                {
+                    // if only one block of data was sent for this message
+                    if (receiveStream.Length == 0)
+                    {
+                        // if the message is empty, skip to the next message
+                        if (result.Count == 0)
+                            continue;
+                        // read the message
+                        bufferStream.ResetLength(result.Count);
+                        var message = await _serializer.ReadAsync<OperationMessage>(bufferStream, RequestAborted);
+                        // dispatch the message
+                        if (message != null)
+                            await OnDispatchMessageAsync(operationMessageProcessor, message);
+                    }
+                    else
+                    {
+                        // if there is any data in this block, add it to the buffer
+                        if (result.Count > 0)
+                            receiveStream.Write(buffer, 0, result.Count);
+                        // read the message from the buffer
+                        receiveStream.Position = 0;
+                        var message = await _serializer.ReadAsync<OperationMessage>(receiveStream, RequestAborted);
+                        // clear the buffer
+                        receiveStream.SetLength(0);
+                        // dispatch the message
+                        if (message != null)
+                            await OnDispatchMessageAsync(operationMessageProcessor, message);
+                    }
+                }
+                else
+                {
+                    // if there is any data in this block, add it to the buffer
+                    if (result.Count > 0)
+                        receiveStream.Write(buffer, 0, result.Count);
+                }
+            }
+        }
+        catch (WebSocketException)
+        {
+        }
+        finally
+        {
+            // prevent any more messages from being sent
+            _outputClosed.TrySetResult(false);
+            // prevent any more messages from attempting to send
+            // note: this statement should be redundant, as WebSocketHandler should dispose operationMessageProcessor
+            operationMessageProcessor.Dispose();
+        }
+    }
+
+    /// <inheritdoc/>
+    public Task CloseAsync()
+        => CloseAsync(1000, null);
+
+    /// <inheritdoc/>
+    public Task CloseAsync(int eventId, string? description)
```

Comment:
Worth to add a comment to `IWebSocketConnection` that close methods do not really close connection immediately.